File: src/DashSpec.Core/Parsing/SqlDialectParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class SqlDialectParser
{
    public static SqlDialect Parse(string raw)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(raw);

        return raw.Trim().ToLowerInvariant() switch
        {
            "tsql" or "mssql" or "sqlserver" => SqlDialect.TSql,
            "postgres" or "postgresql" or "pg" => SqlDialect.Postgres,
            "generic" => SqlDialect.Generic,
            _ => throw new DashSpecParseException(
                $"Unknown @sqldialect '{raw}'. Expected: tsql, postgres, generic."),
        };
    }
}

File: src/DashSpec.Modeling.Parse/Document/SqlDialectParser.fs
namespace DashSpec.Modeling.Parse.Document

open System
open DashSpec.Modeling.Core

module SqlDialectParser =

    let parse (raw: string) =
        if String.IsNullOrWhiteSpace raw then
            invalidArg "raw" "SQL dialect value is required."

        match raw.Trim().ToLowerInvariant() with
        | "tsql" | "mssql" | "sqlserver" -> SqlDialect.TSql
        | "postgres" | "postgresql" | "pg" -> SqlDialect.Postgres
        | "generic" -> SqlDialect.Generic
        | _ ->
            raise (DashSpecParseException($"Unknown @sqldialect '{raw}'. Expected: tsql, postgres, generic."))
